events/janus_mqttevh.c

```diff
@@ -0,0 +1,897 @@
+/*! \file   janus_mqttevh.c
+ *
+ * \author Copyright Olle E. Johansson <oej@edvina.net>
+ *
+ * Based on the mqtt transport by
+ * \author Andrei Nesterov <ae.nesterov@gmail.com>
+ * and the RabbitMQ event plugin by
+ * \author Piter Konstantinov <pit.here@gmail.com>
+ *
+ * \copyright GNU General Public License v3
+ * \brief Janus MQTT transport plugin
+ *
+ * \ingroup eventhandlers
+ * \ref eventhandlers
+ *
+ * \todo Add will handling
+ * \todo Add TLS support (not SSL)
+ * \todo Add support for addplugin in topic
+ * \todo Improve the MQTT connect message (the first one)
+ * \todo Fix random client id if not configured
+ */
+
+#include "eventhandler.h"
+
+/* We use the Paho MQTT library  
+ * http://www.eclipse.org/paho/clients/c/
+ */
+#include <MQTTAsync.h>
+
+#include "../debug.h"
+#include "../config.h"
+#include "../utils.h"
+#include "../events.h"
+
+/* Plugin information */
+#define JANUS_MQTTEVH_VERSION               1
+#define JANUS_MQTTEVH_VERSION_STRING        "0.1.0"
+#define JANUS_MQTTEVH_DESCRIPTION           "An MQTT event handler plugin for Janus."
+#define JANUS_MQTTEVH_NAME                  "JANUS MqttEventHandler plugin"
+#define JANUS_MQTTEVH_AUTHOR                "Olle E. Johansson, Edvina AB"
+#define JANUS_MQTTEVH_PACKAGE               "janus.eventhandler.mqttevh"
+
+/* Plugin methods */
+janus_eventhandler *create(void);
+static int janus_mqttevh_init(const char *config_path);
+static void janus_mqttevh_destroy(void);
+static int janus_mqttevh_get_api_compatibility(void);
+static int janus_mqttevh_get_version(void);
+static const char *janus_mqttevh_get_version_string(void);
+static const char *janus_mqttevh_get_description(void);
+static const char *janus_mqttevh_get_name(void);
+static const char *janus_mqttevh_get_author(void);
+static const char *janus_mqttevh_get_package(void);
+static void janus_mqttevh_incoming_event(json_t *event);
+
+static int janus_mqttevh_send_message(void *context, const char *topic, json_t *message);
+static void *janus_mqttevh_handler(void *data);
+
+
+/* Event handler setup */
+static janus_eventhandler janus_mqttevh =
+        JANUS_EVENTHANDLER_INIT (
+                .init = janus_mqttevh_init,
+                .destroy = janus_mqttevh_destroy,
+
+                .get_api_compatibility = janus_mqttevh_get_api_compatibility,
+                .get_version = janus_mqttevh_get_version,
+                .get_version_string = janus_mqttevh_get_version_string,
+                .get_description = janus_mqttevh_get_description,
+                .get_name = janus_mqttevh_get_name,
+                .get_author = janus_mqttevh_get_author,
+                .get_package = janus_mqttevh_get_package,
+
+                .incoming_event = janus_mqttevh_incoming_event,
```

Comment:
As anticipated in a previous comment, we now also support a `handle_request` callback for "poking" the handler via Admin API. Anyway, feel free to disregard that, as I can add it myself once we merge. If you probe a handler that doesn't implement it, the Admin API simply returns an error so nothing breaks anyway.